pkg/uhttp/dbcache.go

```diff
@@ -0,0 +1,674 @@
+		return fmt.Errorf("%s", nilConnection)
+	}
+
+	cacheableResponse, err := httputil.DumpResponse(value, true)
+	if err != nil {
+		return err
+	}
+
+	if value.Request != nil {
+		url = getFullUrl(value.Request)
+	}
+
+	err = d.Insert(ctx,
+		key,
+		cacheableResponse,
+		url,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Remove stored keys.
+func (d *DBCache) Delete(ctx context.Context, key string) error {
+	if d.IsNilConnection() {
+		return fmt.Errorf("%s", nilConnection)
+	}
+
+	err := d.Remove(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *DBCache) cleanup(ctx context.Context) error {
+	if d.IsNilConnection() {
+		return fmt.Errorf("%s", nilConnection)
+	}
+
+	l := ctxzap.Extract(ctx)
+	stats, err := d.getStats(ctx)
+	if err != nil {
+		l.Debug("error getting stats", zap.Error(err))
+		return err
+	}
+
+	l.Debug("summary and stats", zap.Any("stats", stats))
+	err = d.close(ctx)
+	if err != nil {
+		l.Debug("error closing db", zap.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+// Insert data into the cache table.
+func (d *DBCache) Insert(ctx context.Context, key string, value any, url string) error {
+	var (
+		bytes []byte
+		err   error
+		ok    bool
+	)
+	if d.IsNilConnection() {
+		return fmt.Errorf("%s", nilConnection)
+	}
+
+	l := ctxzap.Extract(ctx)
+	if bytes, ok = value.([]byte); !ok {
+		bytes, err = json.Marshal(value)
+		if err != nil {
+			l.Debug("error marshaling data", zap.Error(err))
+			return err
+		}
+	}
+
+	if ok, _ := d.Has(ctx, key); !ok {
+		tx, err := d.db.Begin()
+		if err != nil {
+			l.Debug(failStartTransaction, zap.Error(err))
+			return err
+		}
+
+		_, err = tx.ExecContext(ctx, "INSERT INTO http_cache(key, data, expiration, url) values(?, ?, ?, ?)",
+			key,
+			bytes,
+			(time.Now().UnixNano() + d.waitDuration),
+			url,
+		)
+		if err != nil {
+			if errtx := tx.Rollback(); errtx != nil {
+				l.Debug(failRollback, zap.Error(errtx))
+			}
+
+			l.Debug(failInsert, zap.Error(err))
+			return err
+		}
+
+		err = tx.Commit()
+		if err != nil {
+			if errtx := tx.Rollback(); errtx != nil {
+				l.Debug(failRollback, zap.Error(errtx))
+			}
+
+			l.Debug(failInsert, zap.Error(err))
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Has query for cached keys.
+func (d *DBCache) Has(ctx context.Context, key string) (bool, error) {
+	if d.IsNilConnection() {
+		return false, fmt.Errorf("%s", nilConnection)
+	}
+
+	l := ctxzap.Extract(ctx)
+	rows, err := d.db.Query("SELECT data FROM http_cache where key = ?", key)
+	if err != nil {
+		l.Debug("Failed to query cache table", zap.Error(err))
+		return false, err
+	}
+
+	defer rows.Close()
+	for rows.Next() {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// IsNilConnection check if the database connection is nil.
+func (d *DBCache) IsNilConnection() bool {
+	return d.db == nil
+}
+
+// Select query for cached response.
+func (d *DBCache) Select(ctx context.Context, key string) ([]byte, error) {
```

Comment:
This should also be private. Get() is how people get stuff from the cache.

Reply:
ok, I added a new name for this method because it causes an error if I only make it private.

Reply:
Fixed